### ISSUE TITLE
Systematic Asset Loss via Floor Division in TVS Split Operations

### ISSUE THREAD

**Issue by 0xloophole** [dualguard/2026-01-pre-launch-bug-bounty-alignerz#65]
## Summary

The `computeSplitArrays` function in the TVSCalculator library uses floor division when calculating split amounts, resulting in permanent token loss on every split operation. This behavior is not documented and is an artifact of Solidity's integer division semantics.

Because Solidity performs floor division by default, splitting Token Vesting Structure (TVS) NFTs causes rounding down of each split portion independently. The lost tokens remain in the TVSManager contract but are never allocated to any NFT, making them permanently inaccessible. Consequently, users lose real economic value with every split operation, with severity scaling linearly by token price: at $0.10/token this represents negligible loss, but at $10/token users lose $10+ per split, and at $100/token the losses constitute potential legal liability.

## Vulnerable Code

The vulnerability is located in `src/libraries/TVSCalculator.sol`. The split calculation function performs floor division without tracking remainders.

```solidity
// File: src/libraries/TVSCalculator.sol (Line 82)
function computeSplitArrays(TVSManager.Allocation memory allocation, uint256 percentage, uint256 nbOfFlows)
    internal
    pure
    returns (
        uint256[] memory newVestingPeriods,
        uint256[] memory newStartTimes,
        uint256[] memory newAmounts,
        uint256[] memory newClaimedAmounts,
        bool[] memory newClaimedFlows
    )
{
    // ... initialization ...
    
    uint256[] memory baseAmounts = allocation.amounts;
    
    for (uint256 j; j < nbOfFlows;) {
        // VULNERABILITY: Floor division loses tokens on every split
        // (baseAmounts[j] * percentage) / BASIS_POINT rounds down
        // Lost tokens are never tracked or recovered
        uint256 amount = (baseAmounts[j] * percentage) / BASIS_POINT;
        require(amount > 0, Splitting_Should_Not_Zero_Down_Amounts());
        
        // The lost remainder stays in TVSManager but isn't allocated to any NFT
        newAmounts[j] = amount;
        // ...
        unchecked { ++j; }
    }
}
```

This propagates to `TVSManager.splitTVS()` in `src/contracts/vesting/TVSManager.sol`:

```solidity
// File: src/contracts/vesting/TVSManager.sol (Line 180)
function splitTVS(uint256[] calldata percentages, uint256 splitNftId)
    external
    returns (uint256, uint256[] memory)
{
    // ... validation ...
    
    for (uint256 i; i < nbOfTVS;) {
        uint256 nftId = nftContract.mint(msg.sender);
        
        // VULNERABILITY: Calls computeSplitArrays which loses tokens
        (
            allocationOf[nftId].vestingPeriods,
            allocationOf[nftId].vestingStartTimes,
            allocationOf[nftId].amounts,
            allocationOf[nftId].claimedAmounts,
            allocationOf[nftId].claimedFlows
        ) = TVSCalculator.computeSplitArrays(allocation, percentages[i], nbOfFlows);
        
        // Lost tokens remain in TVSManager balance but are not in any NFT allocation
        // At high token prices, this becomes measurable economic loss
        
        unchecked { ++i; }
    }
}
```

## Impact

This issue creates **Price-Dependent Asset Loss** and **Economic Inefficiency**:

1. **Systematic Token Loss**: Each split operation loses 1-5 tokens (depending on flow count) due to floor division rounding. Example: splitting 10,001 tokens 50/50 yields two NFTs with 5,000 tokens each—the 1 lost token remains locked in TVSManager with no recovery mechanism.

2. **Severity Scales with Token Price**: The economic impact scales linearly with token price. At $0.10/token, users lose $0.10 per split (minor). At $10/token, users lose $10+ per split (major). At $100/token, users lose $100+ per split, creating potential legal liability for undisclosed asset loss.

3. **Multi-Flow Amplification**: Users with merged TVS NFTs containing multiple flows lose tokens independently per flow. A 5-flow NFT loses 5 tokens per split: at $10/token this is $50 loss, at $100/token this is $500 loss per operation.

4. **Protocol-Level Cumulative Loss**: Assuming 10,000 users performing 5 splits/year with average 1.5 token loss per split: 75,000 tokens/year lost. At $10/token = $750,000/year in destroyed value. This value benefits no party—it's pure loss, unlike a fee which accrues to the protocol.

5. **Competitive Disadvantage**: Competing DeFi protocols (Uniswap V3, Sablier V2) implement perfect precision with zero loss. Users comparing platforms will prefer competitors, causing market share erosion and liquidity fragmentation.

## Recommended Mitigation

Track remainders during splits and assign to the first split portion to achieve perfect conservation:

```solidity
// File: src/contracts/vesting/TVSManager.sol
function splitTVS(uint256[] calldata percentages, uint256 splitNftId)
    external
    returns (uint256, uint256[] memory)
{
    // ... existing validation ...
    
    uint256[] memory newNftIds = new uint256[](nbOfTVS);
    uint256[] memory totalAllocatedPerFlow = new uint256[](nbOfFlows);
    
    // Create all splits using existing logic
    for (uint256 i; i < nbOfTVS;) {
        uint256 nftId = nftContract.mint(msg.sender);
        newNftIds[i] = nftId;
        
        (
            allocationOf[nftId].vestingPeriods,
            allocationOf[nftId].vestingStartTimes,
            allocationOf[nftId].amounts,
            allocationOf[nftId].claimedAmounts,
            allocationOf[nftId].claimedFlows
        ) = TVSCalculator.computeSplitArrays(allocation, percentages[i], nbOfFlows);
        
        // Track total allocated per flow
        for (uint256 j = 0; j < nbOfFlows; j++) {
            totalAllocatedPerFlow[j] += allocationOf[nftId].amounts[j];
        }
        
        unchecked { ++i; }
    }
    
    // Assign any lost tokens to first NFT (ensures perfect conservation)
    for (uint256 j = 0; j < nbOfFlows; j++) {
        uint256 originalAmount = allocation.amounts[j];
        uint256 totalAllocated = totalAllocatedPerFlow[j];
        
        if (totalAllocated < originalAmount) {
            uint256 loss = originalAmount - totalAllocated;
            allocationOf[newNftIds[0]].amounts[j] += loss;
        }
    }
    
    return (splitNftId, newNftIds);
}
```

## Poc

Add this to AlignerzVestingProtocolTest.t.sol

forge test --match-test test_PoC_AssetLossViaFloorDivision -vv

```solidity
function test_PoC_AssetLossViaFloorDivision() public {
    // 1. Setup a Reward Project with an odd amount
    uint256 oddAmount = 10001; 
    address kol = makeAddr("victimKOL");
    
    vm.startPrank(owner);
    // REMOVE FEES to isolate the rounding bug
    tvsManager.setMergeSplitFeeRates(0, 0); 
    
    alignerz.launchRewardProject(address(token));
    address[] memory kols = new address[](1);
    kols[0] = kol;
    uint256[] memory amounts = new uint256[](1);
    amounts[0] = oddAmount;
    
    token.approve(address(alignerz), oddAmount);
    alignerz.setTVSAllocation(0, oddAmount, 1 days, kols, amounts);
    vm.stopPrank();

    // 2. KOL claims their TVS NFT
    vm.prank(kol);
    uint256 originalNftId = alignerz.claimRewardTVS(0);

    // 3. Perform a 50/50 split (5000 basis points each)
    uint256[] memory percentages = new uint256[](2);
    percentages[0] = 5000; 
    percentages[1] = 5000; 

    vm.prank(kol);
    ( , uint256[] memory newNftIds) = tvsManager.splitTVS(percentages, originalNftId);

    // 4. Verification of Loss
    TVSManager.Allocation memory alloc1 = tvsManager.getAllocationOf(newNftIds[0]);
    TVSManager.Allocation memory alloc2 = tvsManager.getAllocationOf(newNftIds[1]);

    uint256 sumOfAllocations = alloc1.amounts[0] + alloc2.amounts[0];
    uint256 lostTokens = oddAmount - sumOfAllocations;

    console.log("Original Amount:", oddAmount);
    console.log("Sum of Split Amounts:", sumOfAllocations);
    console.log("Total Tokens Lost (Unallocated):", lostTokens);

    // With 0% fees, the loss is exactly 1 token due to integer division
    assertEq(lostTokens, 1, "Should have lost exactly 1 token due to rounding");
}
```

```bash
Ran 1 test for test/POC_ApprovalBypass.t.sol:POC_ApprovalBypass
[PASS] test_PoC_AssetLossViaFloorDivision() (gas: 1070394)
Logs:
  TVSManager deployed at: 0xc7183455a4C133Ae270771860664b6B7ec320bB1
  Set AlignerzTVSManager as minter for AlignerzNFT
  AlignerzVesting deployed at: 0x1d1499e622D69689cdf9004d05Ec547d650Ff211
  Set AlignerzVesting as minter for AlignerzNFT
  Original Amount: 10001
  Sum of Split Amounts: 10000
  Total Tokens Lost (Unallocated): 1

Suite result: ok. 1 passed; 0 failed; 0 skipped; finished in 13.84ms (2.00ms CPU time)

```
